src/os/bluestore/BlueStore.cc

```diff
@@ -7246,8 +7247,25 @@ int BlueStore::_do_alloc_write(
     CompressorRef c;
     bool compressed = false;
     if (wctx->compress &&
-	wi.blob_length > min_alloc_size &&
-	(c = compressor) != nullptr) {
+      wi.blob_length > min_alloc_size) {
+      c = select_option(
+	coll,
+	"compression_algorithm",
+	compressor,
+	[&]() {
+	  string val;
+	  if (coll->pool_opts.get(pool_opts_t::COMPRESSION_ALGORITHM, &val)) {
+	    CompressorRef cp = compressor;
+	    if (!cp || cp->get_type_name() != val) {
+	      cp = Compressor::create(cct, val);
+            }
+	    return boost::optional<CompressorRef>(cp);
+	  }
+	  return boost::optional<CompressorRef>();
+	}
+      );
```

Comment:
The goal of this method is to provide effective selection between pool and store settings. As per our earlier discussion I'm planning to expand it with an ability to select corresponding setting using some filtering on option names ( e.g. bluestore_disable_per_pool_settings = bluestore_compression_algorithm, csum_type). And I think this logic is rather out of collection/pool_opts_t scope. 
Moreover we can optimize the selection procedure to avoid per-collection option lookup/retrieval if per-store settings prevail. Hence we need to access these settings inside the method (via f arg) not provide their values externally.